convert2rhel/unit_tests/actions/actions_test.py

```diff
@@ -0,0 +1,793 @@
+
+    @pytest.mark.parametrize(
+        ("stage_dirs",),
+        (
+            (("bad_deps1",),),
+            (("bad_deps1", "good_deps_failed_actions"),),
+            (("good_deps_failed_actions", "bad_deps1"),),
+            (("good_deps1", "deps_on_1"),),
+            (("deps_on_1", "good_deps_failed_actions"),),
+            (("deps_on_1",),),
+        ),
+    )
+    def test_check_dependencies_fail(self, stage_actions, stage_dirs):
+        stage = None
+        for stage_dir in stage_dirs:
+            stage = actions.Stage(stage_dir, next_stage=stage)
+
+        with pytest.raises(actions.DependencyError):
+            stage.check_dependencies()
+
+    def test_check_dependencies_real_actions(self):
+        """Check the Actions we ship have no broken deps."""
+        pre_ponr_changes = actions.Stage("pre_ponr_changes", "changes")
+        system_checks = actions.Stage("system_checks", "checks", pre_ponr_changes)
+
+        system_checks.check_dependencies()
+
+    #
+    # Test that Stage.run() works as expected
+    #
+    @pytest.mark.parametrize(
+        ("stage_dirs", "expected"),
+        (
+            # Single Stages
+            (
+                ("good_deps1",),
+                (("REALTEST", "SECONDTEST", "THIRDTEST", "FOURTHTEST"), (), ()),
+            ),
+            # Test that all action statuses are categorized correctly
+            (
+                ("all_status_actions",),
+                (
+                    ("SUCCESSTEST", "WARNINGTEST"),
+                    ("ERRORTEST", "OVERRIDABLETEST"),
+                    ("SKIPSINGLETEST", "SKIPMULTIPLETEST"),
+                ),
+            ),
+            # Test that exceptions inside of Actions are detected as errors
+            (
+                ("action_exceptions",),
+                (("SUCCESSTEST",), ("DIVIDEBYZEROTEST", "LOGCRITICALTEST"), ()),
+            ),
+            # Multiple Stages
+            (
+                ("deps_on_1", "good_deps1"),
+                (("REALTEST", "SECONDTEST", "THIRDTEST", "FOURTHTEST", "TESTI", "TESTII"), (), ()),
+            ),
+        ),
+    )
+    def test_run(self, stage_actions, stage_dirs, expected):
+        stage = None
+        for stage_dir in stage_dirs:
+            stage = actions.Stage(stage_dir, next_stage=stage)
+
+        actual = stage.run()
+
+        # Assert that the success, fail, and skipped lists are the same
+        assert sorted(action.id for action in actual.successes) == sorted(expected[0])
+        assert sorted(action.id for action in actual.failures) == sorted(expected[1])
+        assert sorted(action.id for action in actual.skips) == sorted(expected[2])
+
+    def test_stages_cannot_be_run_twice(self, stage_actions):
+        """Test that an Action can only be run once."""
+        stage = actions.Stage("good_deps1")
+        stage.run()
+
+        with pytest.raises(actions.ActionError, match="Stage good_deps1 has already run"):
+            stage.run()
```

Comment:
This test is taking information from the previous one? Not sure if I understood it, but by the name of the stage, it seems to be that since it ran in the `test_run` method, it shouldn't run again here... 

If that's the case, wouldn't be better to create a new stage and try to run it twice?

Reply:
I think what you think this should look like is the way it already is.  This test is isolated from `test_run()`.  We create a new stage on line 308.  Then we run it for the first time on line 309 and for the second time on line 312.  Does that sound like what you were asking for?